feat(binary): Host our own apt repo

kapobajza opened the pull request:
This PR solves part of https://github.com/alexpasmantier/television/issues/246, since hosting our own apt repo is just the first step.

Right now it's a draft, because I put the CD code inside of the [.github/workflows/static.yml](https://github.com/alexpasmantier/television/blob/98a3d7b8547268084e5f54f48eb12774e6969d36/.github/workflows/static.yml) file and this, most-likely, doesn't belong there. Where exactly to put it will need to be discussed, but the requirements are:
- to start executing this code after we generate the `.deb` binary
- and to deploy the static GH pages content after we've executed this code

I've deployed a test version of the `.deb` package to my GH pages and you can actually already install `television` via `apt` like this:

```sh
# download the repository public key and put it in the trusted directory
curl -sS https://kapobajza.github.io/television/apt/repository-key.asc | sudo gpg --dearmor --yes -o /etc/apt/trusted.gpg.d/television.gpg

# put the repository download link inside the apt sources list
echo "deb https://kapobajza.github.io/television/apt stable main" | sudo tee /etc/apt/sources.list.d/television.list

sudo apt update

# install the package 🚀 
sudo apt install television
```

alexpasmantier commented:
Great work!

I'm thinking this can go inside of `.github/workflows/cd.yml` for now as a new job. That way it always gets run after the new release is published.

```yml
jobs:
  create-release:
    ...
  publish-release:
    ...
  packaging:
    [here]
```

What do you think?

kapobajza commented:
> Great work!
> 
> I'm thinking this can go inside of `.github/workflows/cd.yml` for now as a new job. That way it always gets run after the new release is published.
> 
> ```yaml
> jobs:
>   create-release:
>     ...
>   publish-release:
>     ...
>   packaging:
>     [here]
> ```
> 
> What do you think?

Yeah, makes sense. I can put it there. Should I also put the GH pages deployment steps there? After the packaging is done, we need to deploy it to GH pages.


Btw for this to work you would also need to generate a GPG key and put it inside the repo secrets, it's being used [for this action](https://github.com/alexpasmantier/television/blob/6f7ccea3a105694d29a0eab8231b7828f2a46da1/.github/workflows/static.yml#L37-L38). And we'll also need to update the key id that's being used [here](https://github.com/alexpasmantier/television/blob/2a15a799cf657c8ea8337b177ebdb48f657437c4/apt/reprepro/conf/distributions#L5). Instructions on how to generate a GPG key are:
1. Run `gpg --full-generate-key`
2. Select RSA (sign only) with a key size of 4096 bits, and follow the prompts to create your key
3. Run `gpg --armor --export-secret-key`. This is your GPG private key ([`secrets.GPG_PRIVATE_KEY`](https://github.com/alexpasmantier/television/blob/6f7ccea3a105694d29a0eab8231b7828f2a46da1/.github/workflows/static.yml#L37)
4. [`secrets.GPG_PASSPHRASE`](https://github.com/alexpasmantier/television/blob/6f7ccea3a105694d29a0eab8231b7828f2a46da1/.github/workflows/static.yml#L38) is, obviously, the passphrase you entered when generating the key
5. Run `gpg --list-keys --keyid-format long`. The id after `rsa4096/...` is your key id that should be used to [replace the existing one](https://github.com/alexpasmantier/television/blob/2a15a799cf657c8ea8337b177ebdb48f657437c4/apt/reprepro/conf/distributions#L5). For example in my case it's:

```
# The key id is 43BD55100555DBF2
pub   rsa4096/43BD55100555DBF2 2025-06-25 [SC]
      199782ABD367B59116D4447E43BD55100555DBF2
```

And that's it.

Sorry for the longer explanation, you are probably aware or may have done something like this in the past. But in case you didn't, I provided it with details. 😅

alexpasmantier commented:
I feel the way to go should be to have the GH pages deployment in a separate workflow which would be triggered based on certain events, typically once the CD workflow finishes:
```yml
name: Deploy to GH pages
on:
  workflow_run:
    workflows: ["CD"]
    branches: [main]
    types: 
      - completed

jobs:
  ...
```

kapobajza commented:
> @kapobajza Rebased the PR on main, added the GPG public keys to the different files and added the secrets to the repo 🎉
> 
> EDIT: the rebase makes it look like I'm the author of the commits but that's just github being github. ![image](https://private-user-images.githubusercontent.com/47638216/460238467-ab38f22c-4e25-4c0e-b3bd-e2ae23253dac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTExODA0NTAsIm5iZiI6MTc1MTE4MDE1MCwicGF0aCI6Ii80NzYzODIxNi80NjAyMzg0NjctYWIzOGYyMmMtNGUyNS00YzBlLWIzYmQtZTJhZTIzMjUzZGFjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjI5VDA2NTU1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ3YTQ4OGMyNTAzZTE3MWE2MzUxYWJkOTdmY2Y5N2Q4MDVkYjRmYzRiMGZhM2VkMWYwNjJjMmY5ZWNmZjQzMWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5YLCxg2ug1k4AfllE6_4UhPkbZufOb7JEwkWAxIesyc)

Great, thank you for that. 😄

I've moved all the ci/cd code from `static.yml` to `cd.yml`. So the PR is no longer a draft.
Just beware that we won't be able to test this out properly until we make a new release, obviously 😅. Since the `.deb` files that were released previously didn't contain the [additional needed metadata](https://github.com/alexpasmantier/television/blob/0580e1fb1758b2d626f6e0015946c4b4bc3a2f1a/Cargo.toml#L139-L151), [this `reprerpo` command](https://github.com/alexpasmantier/television/blob/6e33212923efc23b5f3b29665718a195d2409916/.github/workflows/cd.yml#L340) will fail.

And I have a question: I wanted to include the packaging for `aarch64` also, but I've noticed that the [target is commented out](https://github.com/alexpasmantier/television/blob/6e33212923efc23b5f3b29665718a195d2409916/.github/workflows/cd.yml#L203-L205). Why is that?

alexpasmantier commented:
> Just beware that we won't be able to test this out properly until we make a new release, obviously 😅. Since the .deb files that were released previously didn't contain the [additional needed metadata](https://github.com/alexpasmantier/television/blob/0580e1fb1758b2d626f6e0015946c4b4bc3a2f1a/Cargo.toml#L139-L151), [this reprerpo command](https://github.com/alexpasmantier/television/blob/6e33212923efc23b5f3b29665718a195d2409916/.github/workflows/cd.yml#L340) will fail.

No problem, we'll see how it goes when we release 0.12 and will iterate from there 👍🏻 

> And I have a question: I wanted to include the packaging for aarch64 also, but I've noticed that the [target is commented out](https://github.com/alexpasmantier/television/blob/6e33212923efc23b5f3b29665718a195d2409916/.github/workflows/cd.yml#L203-L205). Why is that?

I believe I commented that out because compilation would fail on aarch64 because of a C library I was relying on for builtin syntax highlighting (namely oniguruma). But this isn't a dependency anymore so it should be fine to re-add that line in the CD workflow 👍🏻

alexpasmantier commented:
@kapobajza should we merge this or do you want to wait for the changelog thing to be done?

kapobajza commented:
> @kapobajza should we merge this or do you want to wait for the changelog thing to be done?

I've just added the deb packaging for the `aarch64` architecture. The testing I could do is very limited, so I would need to do proper testing once we release the 0.12 version.

But if everything's fine from your side, we can go ahead and merge this. We don't have to wait for the changelog thing, that could be a separate PR.
